modules/carto/src/layers/carto-layer.js

```diff
@@ -16,7 +18,16 @@ export default class CartoLayer extends CompositeLayer {
   updateState({changeFlags}) {
     const {data} = this.props;
     if (changeFlags.dataChanged && data) {
-      this._updateTileJSON();
+      this._updateData();
+    }
+  }
+
+  async _updateData () {
+    try {
+      const tilejson = await this._updateTileJSON();
+      this.props.onDataLoad(tilejson);
```

Comment:
`this.setState({tilejson});` should be moved here, given that other methods in this base class expect the state to exist. This will remove the burden of individual subclasses having to remember setting it.